### PR TITLE
[Merged by Bors] - chore(*): remove `refine'`

### PR DESCRIPTION
**Pull request by grunweg**
Some of these changes are simple replacements, others are more involved.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Parcly-Taxel** [comment]
!bench

File: Mathlib/Algebra/Field/ULift.lean [hunk 1]
```diff
@@ -32,15 +32,15 @@ instance instRatCast [RatCast α] : RatCast (ULift α) where ratCast q := up q
 #align ulift.down_rat_cast ULift.down_ratCast
 
 instance divisionSemiring [DivisionSemiring α] : DivisionSemiring (ULift α) := by
-  refine' down_injective.divisionSemiring down .. <;> intros <;> rfl
+  apply down_injective.divisionSemiring down .. <;> intros <;> rfl
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Field/ULift.lean:35]: And below

File: Mathlib/Algebra/Group/Aut.lean [hunk 2]
```diff
@@ -107,7 +107,7 @@ theorem inv_apply_self (e : MulAut M) (m : M) : e⁻¹ (e m) = m :=
 
 /-- Monoid hom from the group of multiplicative automorphisms to the group of permutations. -/
 def toPerm : MulAut M →* Equiv.Perm M := by
-  refine' { toFun := MulEquiv.toEquiv, ..} <;> intros <;> rfl
+  refine { toFun := MulEquiv.toEquiv, map_one' := ?_, map_mul' := ?_ } <;> intros <;> rfl
```
[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Group/Aut.lean:110]: And below

File: Mathlib/Algebra/PUnitInstances.lean [hunk 3]
```diff
@@ -152,16 +151,20 @@ instance instIsScalarTowerOfSMul [SMul R S] : IsScalarTower R S PUnit :=
   ⟨fun _ _ _ => rfl⟩
 
 instance smulWithZero [Zero R] : SMulWithZero R PUnit := by
-  refine' { PUnit.smul with .. } <;> intros <;> exact Subsingleton.elim _ _
```
Comment (mattrobball) [inline Mathlib/Algebra/PUnitInstances.lean:155]: Again and below

File: Mathlib/Algebra/Star/Free.lean [hunk 4]
```diff
@@ -51,7 +51,8 @@ instance : StarRing (FreeAlgebra R X) where
   star_involutive x := by
     unfold Star.star
     simp only [Function.comp_apply]
-    refine' FreeAlgebra.induction R X _ _ _ _ x
+    refine @induction R _ X (fun x ↦ ((lift R (MulOpposite.op ∘ ι R))
```
Comment (mattrobball) [inline Mathlib/Algebra/Star/Free.lean:54]: What happened here?

File: Mathlib/Algebra/Star/Subalgebra.lean [hunk 5]
```diff
@@ -152,7 +152,10 @@ protected theorem smul_mem {x : A} (hx : x ∈ S) (r : R) : r • x ∈ S :=
 #align star_subalgebra.smul_mem StarSubalgebra.smul_mem
 
 /-- Embedding of a subalgebra into the algebra. -/
-def subtype : S →⋆ₐ[R] A := by refine' { toFun := ((↑) : S → A), .. } <;> intros <;> rfl
+def subtype : S →⋆ₐ[R] A := by
+  refine {
```
Comment (mattrobball) [inline Mathlib/Algebra/Star/Subalgebra.lean:156]: Again

File: Mathlib/AlgebraicGeometry/Morphisms/Basic.lean [hunk 6]
```diff
@@ -555,7 +555,7 @@ theorem AffineTargetMorphismProperty.IsLocal.diagonal_affine_openCover_TFAE
   · exact fun H => ⟨Y.affineCover, inferInstance, H Y.affineCover⟩
   tfae_have 2 → 5
   · rintro ⟨𝒰, h𝒰, H⟩
-    refine' ⟨𝒰, inferInstance, fun _ => Scheme.affineCover _, inferInstance, _⟩
+    refine ⟨𝒰, inferInstance, fun _ ↦ Scheme.affineCover _, inferInstance, ?_⟩
```
Comment (mattrobball) [inline Mathlib/AlgebraicGeometry/Morphisms/Basic.lean:558]: The Unicode arrow is in scope?

File: Mathlib/CategoryTheory/Category/KleisliCat.lean [hunk 7]
```diff
@@ -49,7 +49,7 @@ instance KleisliCat.categoryStruct {m} [Monad.{u, v} m] :
 
 instance KleisliCat.category {m} [Monad.{u, v} m] [LawfulMonad m] : Category (KleisliCat m) := by
   -- Porting note: was
-  -- refine' { id_comp' := _, comp_id' := _, assoc' := _ } <;> intros <;> ext <;> unfold_projs <;>
+  -- refine { id_comp' := _, comp_id' := _, assoc' := _ } <;> intros <;> ext <;> unfold_projs <;>
   --  simp only [(· >=> ·), functor_norm]
   refine { id_comp := ?_, comp_id := ?_, assoc := ?_ } <;> intros <;>
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Category/KleisliCat.lean:54]: Again

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 8]
```diff
@@ -274,7 +274,7 @@ def shiftFunctor (n : S) : DifferentialObject S C ⥤ DifferentialObject S C whe
 @[simps!]
 nonrec def shiftFunctorAdd (m n : S) :
     shiftFunctor C (m + n) ≅ shiftFunctor C m ⋙ shiftFunctor C n := by
-  refine' NatIso.ofComponents (fun X => mkIso (shiftAdd X.obj _ _) _) (fun f => _)
+  refine NatIso.ofComponents (fun X => mkIso (shiftAdd X.obj _ _) ?_) (fun f ↦ ?_)
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/DifferentialObject.lean:277]: Again

File: Mathlib/CategoryTheory/Limits/Presheaf.lean [hunk 9]
```diff
@@ -377,9 +377,9 @@ noncomputable def natIsoOfNatIsoOnRepresentables (L₁ L₂ : (Cᵒᵖ ⥤ Type
     [PreservesColimits L₁] [PreservesColimits L₂] (h : yoneda ⋙ L₁ ≅ yoneda ⋙ L₂) : L₁ ≅ L₂ := by
   apply NatIso.ofComponents _ _
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Limits/Presheaf.lean:378]: Again

File: Mathlib/CategoryTheory/Limits/Shapes/Types.lean [hunk 10]
```diff
@@ -246,8 +246,8 @@ def binaryProductFunctor : Type u ⥤ Type u ⥤ Type u where
 explicit binary product functor given by the product type.
 -/
 noncomputable def binaryProductIsoProd : binaryProductFunctor ≅ (prod.functor : Type u ⥤ _) := by
-  refine' NatIso.ofComponents (fun X => _) (fun _ => _)
-  · refine' NatIso.ofComponents (fun Y => _) (fun _ => _)
+  refine NatIso.ofComponents (fun X ↦ ?_) (fun _ ↦ ?_)
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Limits/Shapes/Types.lean:249]: Again

File: Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean [hunk 11]
```diff
@@ -337,7 +337,7 @@ namespace Localization
 /-- Composition of morphisms in the constructed localized category
 for a morphism property that has left calculus of fractions. -/
 noncomputable def Hom.comp {X Y Z : C} (z₁ : Hom W X Y) (z₂ : Hom W Y Z) : Hom W X Z := by
-  refine' Quot.lift₂ (fun a b => a.comp b) _ _ z₁ z₂
+  refine Quot.lift₂ (fun a b => a.comp b) ?_ ?_ z₁ z₂
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Localization/CalculusOfFractions.lean:340]: Again

File: Mathlib/CategoryTheory/Sites/CompatibleSheafification.lean [hunk 12]
```diff
@@ -91,10 +91,10 @@ the sheafification of `P ⋙ F`, functorially in `P`. -/
 noncomputable def sheafificationWhiskerRightIso :
     J.sheafification D ⋙ (whiskeringRight _ _ _).obj F ≅
       (whiskeringRight _ _ _).obj F ⋙ J.sheafification E := by
-  refine' Functor.associator _ _ _ ≪≫ _
-  refine' isoWhiskerLeft (J.plusFunctor D) (J.plusFunctorWhiskerRightIso _) ≪≫ _
-  refine' _ ≪≫ Functor.associator _ _ _
-  refine' (Functor.associator _ _ _).symm ≪≫ _
+  refine Functor.associator _ _ _ ≪≫ ?_
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Sites/CompatibleSheafification.lean:94]: Again

File: Mathlib/CategoryTheory/Sites/DenseSubsite.lean [hunk 13]
```diff
@@ -345,11 +344,11 @@ noncomputable def sheafCoyonedaHom (α : G.op ⋙ ℱ ⟶ G.op ⋙ ℱ'.val) :
 noncomputable def sheafYonedaHom (α : G.op ⋙ ℱ ⟶ G.op ⋙ ℱ'.val) :
     ℱ ⋙ yoneda ⟶ ℱ'.val ⋙ yoneda := by
   let α := sheafCoyonedaHom α
-  refine'
-    { app := _
-      naturality := _ }
+  refine
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Sites/DenseSubsite.lean:347]: Again

File: Mathlib/CategoryTheory/Triangulated/Opposite.lean [hunk 14]
```diff
@@ -255,7 +255,7 @@ equivalence `triangleOpEquivalence C : (Triangle C)ᵒᵖ ≌ Triangle Cᵒᵖ`
 @[simps!]
 noncomputable def counitIso : inverse C ⋙ functor C ≅ 𝟭 _ :=
   NatIso.ofComponents (fun T => by
-    refine' Triangle.isoMk _ _ (Iso.refl _) (Iso.refl _) (Iso.refl _) _ _ _
+    refine Triangle.isoMk _ _ (Iso.refl _) (Iso.refl _) (Iso.refl _) ?_ ?_ ?_
```
Comment (mattrobball) [inline Mathlib/CategoryTheory/Triangulated/Opposite.lean:258]: Again

File: Mathlib/Control/Applicative.lean [hunk 15]
```diff
@@ -71,7 +71,7 @@ end Lemmas
 -- Porting note: we have a monad instance for `Id` but not `id`, mathport can't tell
 -- which one is intended
 
-instance : CommApplicative Id := by refine' { .. } <;> intros <;> rfl
+instance : CommApplicative Id := by constructor; intros; rfl
```
Comment (mattrobball) [inline Mathlib/Control/Applicative.lean:74]: Again

File: Mathlib/Control/Applicative.lean [hunk 16]
```diff
@@ -130,7 +130,7 @@ open CommApplicative
 
 instance {f : Type u → Type w} {g : Type v → Type u} [Applicative f] [Applicative g]
     [CommApplicative f] [CommApplicative g] : CommApplicative (Comp f g) := by
-  refine' { @instLawfulApplicativeComp f g _ _ _ _ with .. }
+  refine { @instLawfulApplicativeComp f g _ _ _ _ with commutative_prod := ?_ }
```
Comment (mattrobball) [inline Mathlib/Control/Applicative.lean:133]: Again

File: Mathlib/Control/Applicative.lean [hunk 17]
```diff
@@ -163,11 +163,11 @@ instance {α} [One α] [Mul α] : Applicative (Const α) where
 -- Also, `simp` didn't close `refl` goals.
 
 instance {α} [Monoid α] : LawfulApplicative (Const α) := by
-  refine' { .. } <;> intros <;> simp [mul_assoc, (· <$> ·), Seq.seq, pure] <;> rfl
```
Comment (mattrobball) [inline Mathlib/Control/Applicative.lean:166]: Again and below

File: Mathlib/Control/Bifunctor.lean [hunk 18]
```diff
@@ -117,29 +117,27 @@ open Functor
 instance Prod.bifunctor : Bifunctor Prod where bimap := @Prod.map
 #align prod.bifunctor Prod.bifunctor
 
-instance Prod.lawfulBifunctor : LawfulBifunctor Prod := by
-  refine' { .. } <;> intros <;> rfl
+instance Prod.lawfulBifunctor : LawfulBifunctor Prod := by constructor <;> intros <;> rfl
```
Comment (mattrobball) [inline Mathlib/Control/Bifunctor.lean:120]: Again and below

File: Mathlib/Control/Traversable/Basic.lean [hunk 19]
```diff
@@ -265,7 +265,7 @@ class LawfulTraversable (t : Type u → Type u) [Traversable t] extends LawfulFu
 instance : Traversable Id :=
   ⟨id⟩
 
-instance : LawfulTraversable Id := by refine' { .. } <;> intros <;> rfl
+instance : LawfulTraversable Id := by constructor <;> intros <;> rfl
```
Comment (mattrobball) [inline Mathlib/Control/Traversable/Basic.lean:268]: Again

File: Mathlib/Data/DList/Instances.lean [hunk 20]
```diff
@@ -26,9 +26,7 @@ variable (α : Type*)
 /-- The natural equivalence between lists and difference lists, using
 `DList.ofList` and `DList.toList`. -/
 def DList.listEquivDList : List α ≃ DList α := by
-  refine'
-      { toFun := DList.ofList
-        invFun := DList.toList.. } <;>
+  refine { toFun := DList.ofList, invFun := DList.toList, left_inv := ?_, right_inv := ?_ } <;>
```
Comment (mattrobball) [inline Mathlib/Data/DList/Instances.lean:29]: Again

File: Mathlib/Data/Finset/Basic.lean [hunk 21]
```diff
@@ -1318,9 +1318,11 @@ lemma Nonempty.exists_cons_eq (hs : s.Nonempty) : ∃ t a ha, cons a t ha = s :=
 /-- Inserting an element to a finite set is equivalent to the option type. -/
 def subtypeInsertEquivOption {t : Finset α} {x : α} (h : x ∉ t) :
     { i // i ∈ insert x t } ≃ Option { i // i ∈ t } := by
-  refine'
```
Comment (mattrobball) [inline Mathlib/Data/Finset/Basic.lean:1321]: Again

File: Mathlib/Data/Fintype/Quotient.lean [hunk 22]
```diff
@@ -29,9 +29,9 @@ def Quotient.finChoiceAux {ι : Type*} [DecidableEq ι] {α : ι → Type*} [S :
     ∀ l : List ι, (∀ i ∈ l, Quotient (S i)) → @Quotient (∀ i ∈ l, α i) (by infer_instance)
   | [], _ => ⟦fun i h => nomatch List.not_mem_nil _ h⟧
   | i :: l, f => by
-    refine'
+    refine
```
Comment (mattrobball) [inline Mathlib/Data/Fintype/Quotient.lean:32]: Again

File: Mathlib/Data/Multiset/Functor.lean [hunk 23]
```diff
@@ -27,7 +27,7 @@ theorem fmap_def {α' β'} {s : Multiset α'} (f : α' → β') : f <$> s = s.ma
   rfl
 #align multiset.fmap_def Multiset.fmap_def
 
-instance : LawfulFunctor Multiset := by refine' { .. } <;> intros <;> (try simp); rfl
+instance : LawfulFunctor Multiset := by constructor <;> intros <;> (try simp); rfl
```
Comment (mattrobball) [inline Mathlib/Data/Multiset/Functor.lean:30]: Again

File: Mathlib/Data/Num/Lemmas.lean [hunk 24]
```diff
@@ -410,11 +410,14 @@ instance addMonoidWithOne : AddMonoidWithOne Num :=
 #align num.add_monoid_with_one Num.addMonoidWithOne
 
 instance commSemiring : CommSemiring Num := by
-  refine'
+  refine
```
Comment (mattrobball) [inline Mathlib/Data/Num/Lemmas.lean:413]: Again

Comment (mattrobball) [reply Mathlib/Data/Num/Lemmas.lean:413]: And below

File: Mathlib/GroupTheory/Sylow.lean [hunk 25]
```diff
@@ -466,11 +466,9 @@ theorem Sylow.normalizer_sup_eq_top {p : ℕ} [Fact p.Prime] {N : Subgroup G} [N
   apply mul_mem_sup (N.inv_mem n.2)
   rw [Sylow.smul_def, ← mul_smul, ← MulAut.conjNormal_val, ← MulAut.conjNormal.map_mul,
     Sylow.ext_iff, Sylow.pointwise_smul_def, Subgroup.pointwise_smul_def] at hn
-  refine' fun x =>
-    (mem_map_iff_mem
-            (show Function.Injective (MulAut.conj (↑n * g)).toMonoidHom from
-              (MulAut.conj (↑n * g)).injective)).symm.trans
-      _
+  convert fun x ↦ (mem_map_iff_mem
```
Comment (mattrobball) [inline Mathlib/GroupTheory/Sylow.lean:469]: These play the loose games with meta variable assignments

File: Mathlib/LinearAlgebra/InvariantBasisNumber.lean [hunk 26]
```diff
@@ -276,9 +276,10 @@ private def induced_map (I : Ideal R) (e : (ι → R) →ₗ[R] ι' → R) :
     `R^n/I^n ≃ R^m/I^m`. -/
 private def induced_equiv [Fintype ι'] (I : Ideal R) (e : (ι → R) ≃ₗ[R] ι' → R) :
     ((ι → R) ⧸ I.pi ι) ≃ₗ[R ⧸ I] (ι' → R) ⧸ I.pi ι' := by
-  refine'
+  refine
```
Comment (mattrobball) [inline Mathlib/LinearAlgebra/InvariantBasisNumber.lean:279]: Again

File: Mathlib/LinearAlgebra/Matrix/ToLinearEquiv.lean [hunk 27]
```diff
@@ -82,12 +82,12 @@ See `Matrix.toLinearEquiv'` for this result on `n → R`.
 @[simps apply]
 noncomputable def toLinearEquiv [DecidableEq n] (A : Matrix n n R) (hA : IsUnit A.det) :
     M ≃ₗ[R] M := by
-  refine'
+  refine
```
Comment (mattrobball) [inline Mathlib/LinearAlgebra/Matrix/ToLinearEquiv.lean:85]: Again

Comment (mattrobball) [reply Mathlib/LinearAlgebra/Matrix/ToLinearEquiv.lean:85]: And below

File: Mathlib/MeasureTheory/MeasurableSpace/Basic.lean [hunk 28]
```diff
@@ -1753,7 +1753,7 @@ def piCongrRight (e : ∀ a, π a ≃ᵐ π' a) : (∀ a, π a) ≃ᵐ ∀ a, π
 variable (π) in
 /-- Moving a dependent type along an equivalence of coordinates, as a measurable equivalence. -/
 def piCongrLeft (f : δ ≃ δ') : (∀ b, π (f b)) ≃ᵐ ∀ a, π a := by
-  refine' { Equiv.piCongrLeft π f with .. }
```
Comment (mattrobball) [inline Mathlib/MeasureTheory/MeasurableSpace/Basic.lean:1756]: Again

Comment (mattrobball) [reply Mathlib/MeasureTheory/MeasurableSpace/Basic.lean:1756]: And below

File: Mathlib/NumberTheory/Dioph.lean [hunk 29]
```diff
@@ -190,14 +190,18 @@ theorem mul_apply (f g : Poly α) (x : α → ℕ) : (f * g) x = f x * g x := rf
...
-  refine' { add := ((· + ·) : Poly α → Poly α → Poly α)
-            neg := (Neg.neg : Poly α → Poly α)
-            sub := Sub.sub
-            zero := 0
-            nsmul := @nsmulRec _ ⟨(0 : Poly α)⟩ ⟨(· + ·)⟩
-            zsmul := @zsmulRec _ ⟨(0 : Poly α)⟩ ⟨(· + ·)⟩ ⟨Neg.neg⟩
-              (@nsmulRec _ ⟨(0 : Poly α)⟩ ⟨(· + ·)⟩)
-            .. }
+  refine {
+    add := ((· + ·) : Poly α → Poly α → Poly α)
```
Comment (mattrobball) [inline Mathlib/NumberTheory/Dioph.lean:194]: Again

File: Mathlib/Order/BooleanAlgebra.lean [hunk 30]
```diff
@@ -873,10 +873,11 @@ protected abbrev Function.Injective.booleanAlgebra [Sup α] [Inf α] [Top α] [B
 
 end lift
 
-instance PUnit.instBooleanAlgebra : BooleanAlgebra PUnit := by
-  refine'
+instance PUnit.instBooleanAlgebra : BooleanAlgebra PUnit :=
```
Comment (mattrobball) [inline Mathlib/Order/BooleanAlgebra.lean:876]: Again

File: Mathlib/Order/CompleteBooleanAlgebra.lean [hunk 31]
```diff
@@ -590,11 +590,11 @@ namespace PUnit
 variable (s : Set PUnit.{u + 1}) (x y : PUnit.{u + 1})
 
 instance instCompleteAtomicBooleanAlgebra : CompleteAtomicBooleanAlgebra PUnit := by
-  refine'
```
Comment (mattrobball) [inline Mathlib/Order/CompleteBooleanAlgebra.lean:593]: Again

File: Mathlib/Order/RelIso/Basic.lean [hunk 32]
```diff
@@ -503,10 +503,10 @@ theorem ofMapRelIff_coe (f : α → β) [IsAntisymm α r] [IsRefl β s]
 def ofMonotone [IsTrichotomous α r] [IsAsymm β s] (f : α → β) (H : ∀ a b, r a b → s (f a) (f b)) :
     r ↪r s := by
   haveI := @IsAsymm.isIrrefl β s _
-  refine' ⟨⟨f, fun a b e => _⟩, @fun a b => ⟨fun h => _, H _ _⟩⟩
-  · refine' ((@trichotomous _ r _ a b).resolve_left _).resolve_right _ <;>
+  refine ⟨⟨f, fun a b e ↦ ?_⟩, @fun a b => ⟨fun h ↦ ?_, H _ _⟩⟩
```
Comment (mattrobball) [inline Mathlib/Order/RelIso/Basic.lean:506]: Again?

**Review by mattrobball — COMMENTED** [review]
Made it up to `RingTheory`

File: Mathlib/CategoryTheory/Sites/CompatibleSheafification.lean (hunk 12 above)
Comment (Parcly-Taxel) [reply Mathlib/CategoryTheory/Sites/CompatibleSheafification.lean:94]: I didn't change arrows in at least _some_ of these files…

File: Mathlib/Algebra/Field/ULift.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Algebra/Field/ULift.lean:35]: I don't know how to fix this instance!

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Field/ULift.lean:35]:
```lean
instance divisionSemiring [DivisionSemiring α] : DivisionSemiring (ULift α) where 
  toSemiring := ULift.semiring
  toNNRatCast := ULift.instNNRatCast
  __ := ULift.groupWithZero
  nnqsmul := (· • ·) 
  nnqsmul_def := fun _ x => congrArg up <| DivisionSemiring.nnqsmul_def _ x.down 
  nnratCast_def := fun _ => congrArg up <| DivisionSemiring.nnratCast_def _
```
with importing `Mathlib.Algebra.Ring.ULift`

[the remark this answers is not in the document]
Comment (mattrobball) [reply Mathlib/Algebra/Field/ULift.lean:35]: Another solution is to add default values for `Function.Injective.divisionSemiring` 
```lean
protected abbrev divisionSemiring [DivisionSemiring β] (zero : f 0 = 0 := rfl) (one : f 1 = 1 := rofl)
    (add : ∀ x y, f (x + y) = f x + f y := fun _ _ => rfl) (mul : ∀ x y, f (x * y) = f x * f y := fun _ _ => rfl)
    (inv : ∀ x, f x⁻¹ = (f x)⁻¹ := fun _ => rfl) (div : ∀ x y, f (x / y) = f x / f y := fun _ _ => rfl)
    (nsmul : ∀ (n : ℕ) (x), f (n • x) = n • f x := fun _ _ => rfl) (nnqsmul : ∀ (q : ℚ≥0) (x), f (q • x) = q • f x := fun _ _ => rfl)
    (npow : ∀ (x) (n : ℕ), f (x ^ n) = f x ^ n := fun _ _ => rfl) (zpow : ∀ (x) (n : ℤ), f (x ^ n) = f x ^ n := fun _ _ => rfl)
    (natCast : ∀ n : ℕ, f n = n := fun _ => rfl) (nnratCast : ∀ q : ℚ≥0, f q = q := fun _ => rfl) : DivisionSemiring α where
```

[the remark this answers is not in the document]
Comment (Parcly-Taxel) [reply Mathlib/Algebra/Field/ULift.lean]: OK, I have followed your solution for lack of a better one (in terms of using another import).

File: Mathlib/Algebra/Star/Free.lean (hunk 4 above)
Comment (Parcly-Taxel) [reply Mathlib/Algebra/Star/Free.lean:54]: Nothing happened, I have resolved all your comments.
```lean4
let y := lift R (X := X) (MulOpposite.op ∘ ι R)
    apply induction (C := fun x ↦ (y (y x).unop).unop = x) ?_ ?_ ?_ ?_ x
```

**Review by Parcly-Taxel — APPROVED** [review]

**Comment by grunweg** [comment]
!bench

File: Mathlib/Algebra/Ring/Hom/Defs.lean [hunk 33]
```diff
@@ -195,8 +195,8 @@ end
 variable [NonUnitalNonAssocSemiring α] [NonUnitalNonAssocSemiring β]
 
 /-- The identity non-unital ring homomorphism from a non-unital semiring to itself. -/
-protected def id (α : Type*) [NonUnitalNonAssocSemiring α] : α →ₙ+* α := by
-  refine' { toFun := id.. } <;> intros <;> rfl
+protected def id (α : Type*) [NonUnitalNonAssocSemiring α] : α →ₙ+* α where
+  toFun := id; map_mul' _ _ := rfl; map_zero' := rfl; map_add' _ _ := rfl
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Hom/Defs.lean:199]:
```suggestion
  toFun := id
  map_mul' _ _ := rfl
  map_zero' := rfl
  map_add' _ _ := rfl
```

File: Mathlib/Algebra/Ring/Hom/Defs.lean [hunk 34]
```diff
@@ -626,8 +626,8 @@ def mk' [NonAssocSemiring α] [NonAssocRing β] (f : α →* β)
 variable {_ : NonAssocSemiring α} {_ : NonAssocSemiring β}
 
 /-- The identity ring homomorphism from a semiring to itself. -/
-def id (α : Type*) [NonAssocSemiring α] : α →+* α := by
-  refine' { toFun := _root_.id.. } <;> intros <;> rfl
+def id (α : Type*) [NonAssocSemiring α] : α →+* α where
+  toFun := _root_.id; map_zero' := rfl; map_one' := rfl; map_add' _ _ := rfl; map_mul' _ _ := rfl
```
Comment (jcommelin) [inline Mathlib/Algebra/Ring/Hom/Defs.lean:630]:
```suggestion
  toFun := _root_.id
  map_zero' := rfl
  map_one' := rfl
  map_add' _ _ := rfl
  map_mul' _ _ := rfl
```

**Review by jcommelin — APPROVED** [review]
I would be happy with merging this. But I also wouldn't mind having a 2nd opinion.

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**